pkg/cloud/azurestack/azurestack_test.go

```diff
@@ -62,3 +72,89 @@ func TestResourcesRenderingSmoke(t *testing.T) {
 		})
 	}
 }
+
+func makeInfrastructureResource(platform configv1.PlatformType, cloudName configv1.AzureCloudEnvironment) *configv1.Infrastructure {
+	cfg := configv1.Infrastructure{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "cluster",
+		},
+		Status: configv1.InfrastructureStatus{
+			PlatformStatus: &configv1.PlatformStatus{
+				Type: platform,
+			},
+		},
+		Spec: configv1.InfrastructureSpec{
+			CloudConfig: configv1.ConfigMapFileReference{
+				Name: infraCloudConfName,
+				Key:  infraCloudConfKey,
+			},
+			PlatformSpec: configv1.PlatformSpec{
+				Type: platform,
+			},
+		},
+	}
+
+	if platform == configv1.AzurePlatformType {
+		cfg.Status.PlatformStatus.Azure = &configv1.AzurePlatformStatus{
+			CloudName: cloudName,
+		}
+	}
+
+	return &cfg
+}
+
+// This test is a little complicated with all the JSON marshalling and
+// unmarshalling, but it is necessary due to the nature of how this data
+// is stored in Kuberenetes. The ConfigMaps containing the cloud config
+// will have string encoded JSON objects in them, due to the non-deterministic
+// natue of map object in Go we will need to examine the data instead of
+// comparing strings.
+func TestCloudConfigTransformer(t *testing.T) {
+	tc := []struct {
+		name     string
+		source   azure.Config
+		expected azure.Config
+		infra    *configv1.Infrastructure
+		errMsg   string
+	}{
+		{
+			name:     "Azure Stack Hub sets the vmType to standard",
+			source:   azure.Config{},
+			expected: azure.Config{VMType: "standard"},
+			infra:    makeInfrastructureResource(configv1.AzurePlatformType, configv1.AzureStackCloud),
+		},
+		{
+			name:     "Azure Stack Hub doesn't modify vmType if user set",
+			source:   azure.Config{VMType: "vmss"},
+			expected: azure.Config{VMType: "vmss"},
+			infra:    makeInfrastructureResource(configv1.AzurePlatformType, configv1.AzureStackCloud),
+		},
+		{
+			name:   "Non Azure Stack Hub returns an error",
+			source: azure.Config{},
+			infra:  makeInfrastructureResource(configv1.AzurePlatformType, configv1.AzurePublicCloud),
+			errMsg: fmt.Sprintf("invalid platform, expected CloudName to be %s", configv1.AzureStackCloud),
+		},
+	}
+
+	for _, tc := range tc {
+		t.Run(tc.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			src, err := json.Marshal(tc.source)
+			g.Expect(err).NotTo(HaveOccurred(), "Marshal of source data should succeed")
+
+			actual, err := CloudConfigTransformer(string(src), tc.infra, nil)
+			if tc.errMsg != "" {
+				g.Expect(err).Should(MatchError(tc.errMsg))
+				g.Expect(actual).Should(Equal(""))
+			} else {
+				var observed azure.Config
+				err = json.Unmarshal([]byte(actual), &observed)
+				g.Expect(err).NotTo(HaveOccurred(), "Unmarshal of observed data should succeed")
```

Comment:
Nit, could inline this
```suggestion
				g.Expect(json.Unmarshal([]byte(actual), &observed)).To(Succeed(), "Unmarshal of observed data should succeed")
```